File: apps/worker/src/better_answers_worker/contract_digest.py
import hashlib
from pathlib import Path

CONTRACTS_ROOT = Path(__file__).resolve().parents[4] / "contracts"

CONTRACT_STAMP_MODULE = Path(__file__).resolve().parent / "contract_stamp.py"

# It carries no agreement, and a typo fix in prose must not idle this worker.
OUTSIDE_THE_DIGEST = "README.md"

GENERATE = "cd apps/worker && uv run --frozen generate-contract-stamp"


def contract_files(root: Path) -> list[str]:
    found: list[str] = []
    for file in root.rglob("*"):
        # A link is a path, not content. Said here because `is_file` follows one.
        if file.is_symlink() or not file.is_file():
            continue
        relative = file.relative_to(root)
        if any(part.startswith(".") for part in relative.parts):
            continue
        written = "/".join(relative.parts)
        if written != OUTSIDE_THE_DIGEST:
            found.append(written)
    return sorted(found, key=lambda relative: relative.encode("utf-8"))


# Length-prefixed so a boundary cannot be forged: without it, content holding a
# newline and a plausible path could pose as a second file.
def contract_digest(root: Path) -> str:
    stream = hashlib.sha256()
    for relative in contract_files(root):
        # Bytes, never decoded or newline-normalised: a CRLF checkout is a different
        # digest, and that is right — the tiers would be reading different bytes.
        content = (root / relative).read_bytes()
        stream.update(relative.encode("utf-8"))
        stream.update(b"\n")
        stream.update(str(len(content)).encode("ascii"))
        stream.update(b"\n")
        stream.update(content)
    return stream.hexdigest()


def render_contract_stamp(digest: str) -> str:
    return f'# Generated, never edited: {GENERATE}\n\nCONTRACT_DIGEST = "{digest}"\n'

File: packages/schema/src/contract-digest.ts
import { createHash } from "node:crypto";
import { readdirSync, readFileSync } from "node:fs";
import path from "node:path";

export const CONTRACTS_ROOT = path.resolve(import.meta.dirname, "../../../contracts");

export const CONTRACT_STAMP_MODULE = path.resolve(import.meta.dirname, "contract-stamp.ts");

/**
 * Whole and lowercase: a short form in a log is a value somebody compares by eye and gets wrong.
 */
export const CONTRACT_DIGEST_PATTERN = /^[0-9a-f]{64}$/u;

/** It carries no agreement, and a typo fix in prose must not idle a worker. */
const OUTSIDE_THE_DIGEST = "README.md";

const NEWLINE = Buffer.from([0x0a]);

/**
 * JavaScript orders strings by UTF-16 code unit, which ranks an astral path where a byte
 * sort does not.
 */
const byByte = (a: string, b: string): number =>
  Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));

/**
 * Paths relative to `root`, `/`-separated and sorted by UTF-8 bytes. Leaves out links, any path
 * with a segment starting `.`, and the top-level README.
 */
export const contractFiles = (root: string): readonly string[] =>
  readdirSync(root, { recursive: true, withFileTypes: true })
    // A link is a path, not content, and `isFile` is already false for one.
    .filter((entry) => entry.isFile())
    .map((entry) =>
      path.relative(root, path.join(entry.parentPath, entry.name)).split(path.sep).join("/"),
    )
    .filter((relative) => !relative.split("/").some((segment) => segment.startsWith(".")))
    .filter((relative) => relative !== OUTSIDE_THE_DIGEST)
    .toSorted(byByte);

/**
 * A lowercase hex SHA-256 over each contract file's path, byte length and bytes, in turn.
 * Length-prefixed so a boundary cannot be forged: without it, content holding a newline and
 * a plausible path could pose as a second file.
 */
export const contractDigest = (root: string): string => {
  const stream = createHash("sha256");
  for (const relative of contractFiles(root)) {
    /**
     * Bytes, never decoded or newline-normalised: a CRLF checkout is a different digest, and
     * that is right — the tiers would be reading different bytes.
     */
    const content = readFileSync(path.join(root, relative));
    stream.update(Buffer.from(relative, "utf8"));
    stream.update(NEWLINE);
    stream.update(Buffer.from(String(content.length), "ascii"));
    stream.update(NEWLINE);
    stream.update(content);
  }
  return stream.digest("hex");
};

const GENERATE = "pnpm --filter @better-answers/schema run generate:contract-stamp";

export const renderContractStamp = (digest: string): string =>
  [
    `// Generated, never edited: ${GENERATE}`,
    "",
    `export const CONTRACT_DIGEST = "${digest}";`,
    "",
  ].join("\n");
